### PR TITLE
Order PlaceLocations by quality during planning

### PR DESCRIPTION
**Pull request by mvieth**
### Description

The quality field was added by https://github.com/ros-planning/moveit_msgs/commit/cdb9e4aa964269f1a159a8f35c7b634365e86038.
The changes are adapted from pick.cpp, where the Grasps are also ordered by quality.

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_ros/manipulation/pick_place/src/place.cpp
Comment (felixvd) [inline moveit_ros/manipulation/pick_place/src/place.cpp:322]:
```suggestion
  std::vector<std::size_t> place_locations_order(goal.place_locations.size());
```

**Review by felixvd — COMMENTED** [review]
I haven't tested this, but if it works then it is a fine change.

Only suggested to use a longer variable name.

**Review by rhaschke — COMMENTED** [review]
+1 after changing the variable name as suggested by @felixvd.

**Review by felixvd — APPROVED** [review]

**Comment by rhaschke** [comment]
Just a final question before merging: If all qualities are equal (e.g. to zero), the order isn't reversed, right?

**Comment by mvieth** [comment]
> Just a final question before merging: If all qualities are equal (e.g. to zero), the order isn't reversed, right?

The [documentation of std::sort](https://en.cppreference.com/w/cpp/algorithm/sort) says that "The order of equal elements is not guaranteed to be preserved." If you would like that guarantee, `std::stable_sort` could be an option.

File: moveit_ros/manipulation/pick_place/src/place.cpp
Comment (rhaschke) [inline moveit_ros/manipulation/pick_place/src/place.cpp:62]: To preserve order of equal-quality elements:
```suggestion
    if (places_[a].quality > places_[b].quality) return true;
    if (places_[a].quality == places_[b].quality) return a > b;
    return false;
```

File: moveit_ros/manipulation/pick_place/src/place.cpp
Comment (rhaschke) [inline moveit_ros/manipulation/pick_place/src/place.cpp:63]: Sorry, `sort` expects a less operator. The name `OrderPlaceLocationQuality` should reflect this...
What about the name `PlaceLocationLess`?
```suggestion
    if (places_[a].quality == places_[b].quality) return a < b;
```
You should definitely verify that this is the right combination. I just suggested this from the back of my head.

Comment (mvieth) [reply moveit_ros/manipulation/pick_place/src/place.cpp:63]: What about `std::stable_sort` with the original compare operator?

Comment (rhaschke) [reply moveit_ros/manipulation/pick_place/src/place.cpp:63]: `stable_sort` [erzeugt nochmal einen temporären buffer](https://github.com/gcc-mirror/gcc/blob/d9375e490072d1aae73a93949aa158fcd2a27018/libstdc%2B%2B-v3/include/bits/stl_algo.h#L4977)... Daher ist die vorgeschlagene Lösung auf jeden Fall effizienter.

Comment (mvieth) [reply moveit_ros/manipulation/pick_place/src/place.cpp:63]: German :smile: the temporary buffer should be negligible since the sorting algorithm sorts the vector of `std::size_t`, not the `PlaceLocations` itself. Plus using `stable_sort` should be easier to understand than a rather complicated compare logic.

Comment (felixvd) [reply moveit_ros/manipulation/pick_place/src/place.cpp:63]: The cost of that buffer is surely negligible considering the rest of this pipeline. In favor of `stable_sort` (with a comment if necessary).

Also a quick reminder that this is slated to be replaced by MTC, so this can be revisited to your heart's content :0

Comment (rhaschke) [reply moveit_ros/manipulation/pick_place/src/place.cpp:63]: Ok, if there are two votes pro, go for it.

**Review by rhaschke — APPROVED** [review]

File: moveit_ros/manipulation/pick_place/src/place.cpp
Comment (mvieth) [reply moveit_ros/manipulation/pick_place/src/place.cpp]: Should I change `sort` to `stable_sort` in pick.cpp as well?

Comment (rhaschke) [reply moveit_ros/manipulation/pick_place/src/place.cpp]: Makes sense. I wouldn't expect the order to change...

Comment (mvieth) [reply moveit_ros/manipulation/pick_place/src/place.cpp]: Done

**Review by tylerjw — APPROVED** [review]
I read throught the code and it seems like it should work as expected.  I left one small comment asking for a change to the test so it will test the sort operator but won't block merging it as is.

File: moveit_ros/manipulation/pick_place/src/place.cpp
Comment (mvieth) [reply moveit_ros/manipulation/pick_place/src/place.cpp:62]: Hm, this isn't tested for pick, either. I am not sure how this internal sorting could be tested? Currently it is only tested whether the single Grasp/PlaceLocation returns a success. I think there is no way to find out which of several supplied Grasps/PlaceLocations was a success?

**Comment by AndyZe** [comment]
@mvieth I just wanted to reiterate, the pick/place pipeline is probably going to be replaced by MoveIt Task Constructor in the future (especially in ROS2). Might not be worth spending a ton of time on this.

**Comment by v4hn** [comment]
This just got stuck but was approved before. Sorry for the long wait...
Merging.

**Pull request merged**
